src/datacrumbs/common/configuration_manager.cpp

```diff
@@ -0,0 +1,565 @@
+/**
+ * @file configuration_manager.cpp
+ * @brief Implements the ConfigurationManager class for managing DataCrumbs configuration.
+ *
+ * This file contains the implementation of the ConfigurationManager class, which is responsible
+ * for parsing command-line arguments, loading YAML configuration files, and setting up
+ * configuration parameters for the DataCrumbs application. It also includes the ArgumentParser
+ * class for handling command-line arguments and utility functions for deriving and validating
+ * configuration values.
+ */
+
+/**
+ * std headers
+ */
+#include <unistd.h>
+
+#include <algorithm>
+#include <cstring>
+#include <fstream>
+#include <optional>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+/**
+ * Internal headers
+ */
+#include <datacrumbs/common/configuration_manager.h>
+#include <datacrumbs/common/logging.h>  // <-- Added logging header
+#include <datacrumbs/common/singleton.h>
+#include <datacrumbs/common/utils.h>
+/**
+ * External headers
+ */
+#include <yaml-cpp/yaml.h>
+
+namespace datacrumbs {
+
+// Singleton template specialization for ConfigurationManager
+template <>
+std::shared_ptr<datacrumbs::ConfigurationManager>
+    datacrumbs::Singleton<datacrumbs::ConfigurationManager>::instance = nullptr;
+template <>
+bool datacrumbs::Singleton<datacrumbs::ConfigurationManager>::stop_creating_instances = false;
```

Comment:
I'll probably come to it, and this is probably fine, but since function-level statics are initialized deterministically on first call, and automatically initialized atomically once since C++11, this kind of pattern is often now just a function returning a function-level static these days.